pyro/contrib/conjugate/infer.py

```diff
@@ -0,0 +1,164 @@
+
+
+def _compound(base, parent):
+    return _make_cls(base, {}, {"parent": parent})
+
+
+class BetaBinomialPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Beta, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Binomial, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration1 = self._latent.concentration1
+        concentration0 = self._latent.concentration0
+        total_count = self._conditional.total_count
+        reduce_dims = len(obs.size()) - len(concentration1.size())
+        # Unexpand total_count to have the same shape as concentration0.
+        # Raise exception if this isn't possible.
+        total_count = sum_leftmost(total_count, reduce_dims)
+        summed_obs = sum_leftmost(obs, reduce_dims)
+        return dist.Beta(concentration1 + summed_obs,
+                         total_count + concentration0 - summed_obs,
+                         validate_args=self._latent._validate_args)
+
+    def compound(self):
+        return _compound(dist.BetaBinomial, parent=self)(concentration1=self._latent.concentration1,
+                                                         concentration0=self._latent.concentration0,
+                                                         total_count=self._conditional.total_count)
+
+
+class GammaPoissonPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Gamma, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Poisson, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration = self._latent.concentration
+        rate = self._latent.rate
+        reduce_dims = len(obs.size()) - len(rate.size())
+        num_obs = obs.shape[:reduce_dims].numel()
+        summed_obs = sum_leftmost(obs, reduce_dims)
+        return dist.Gamma(concentration + summed_obs, rate + num_obs)
+
+    def compound(self):
+        return _compound(dist.GammaPoisson, parent=self)(concentration=self._latent.concentration,
+                                                         rate=self._latent.rate)
+
+
+class UncollapseConjugateMessenger(ReplayMessenger):
+    r"""
+    Extends `~pyro.poutine.replay_messenger.ReplayMessenger` to uncollapse
+    compound distributions.
+    """
+    def _pyro_sample(self, msg):
+        is_collapsible = getattr(msg["fn"], "collapsible", False)
+        if is_collapsible:
+            observed_node, conj_pair = None, None
+            for site_name in self.trace.observation_nodes:
```

Comment:
Could you also save the conditional `.site_name` and lookup
```py
conditional_name = msg["fn"].parent._conditional.site_name
conditional_node = self.trace[conditional_name]
```

Reply:
Hmm..the issue here is that `msg["fn"].parent._conditional.site_name` might not be set by the time we are at the latent node since that occurs further downstream. We also can't use `self.trace[msg["name"]].parent..` since `msg["name"]` won't be there in `self.trace` as it was collapsed out. 

I guess the only way to retrieve the corresponding pair would be to iterate through the trace and do this round-about check, but we don't need to restrict ourselves to observed nodes.

Reply:
Thanks for explaining.